ouroboros-network/src/Ouroboros/Network/Server/Socket.hs

```diff
@@ -56,11 +60,18 @@ data HandleConnection channel st r where
   Accept :: !st -> !(channel -> IO r) -> HandleConnection channel st r
 
 -- | What to do on a new connection: accept and run this `IO`, or reject.
-type BeginConnection addr channel st r = addr -> st -> STM (HandleConnection channel st r)
+type BeginConnection addr channel st r = Time IO -> addr -> st -> STM (HandleConnection channel st r)
+
+-- | A call back which runs when application starts;
+--
+-- It is needed only because 'BeginConnection' does not have access to the
+-- thread which runs the application.
+--
+type ApplicationStart addr st = addr -> Async () -> st -> STM st
 
 -- | How to update state when a connection finishes. Can use `throwSTM` to
 -- terminate the server.
-type CompleteConnection st r = Either SomeException r -> st -> STM st
+type CompleteConnection addr st r = Result addr r -> st -> STM (st, IO ())
```

Comment:
This is used for two things:
* tracing
* cancelling threads as a result of evaluation of an error policy

It might be cleaner to just return trace points and list of threads, and do that more explicitely.